### ISSUE TITLE
flatten_reflections should wrangle experiment identifiers gracefully

### ISSUE THREAD

**Issue by graeme-winter** [dials/dials#656]
If we have data with identifier eggs and id 0, and another reflection set with identifier spam and id 0, would like it when we flatten these if eggs is id 0 (say) and spam _by magic_ becomes id 1. That way if you have two sets which you put together, you would avoid collisions - when you export these would be unrolled however you like. 

N.B. the numerical id could just as well be mapped to 1702 and 555 to get unique values. 

Currently if you flatten experiments you get a list of experiments, and flattening reflections gives a list with only one _unique_ id in - which makes something like

```
    assert len(experiments) == len(reflections)
```

fail

**Comment by ndevenish** [comment]
I've been looking at this as part of trying to resolve some of the image viewer problems and I somewhat disagree with this approach.

- It's a function that used to literally just do `[x.data for x in reflections]` that now manipulates my data (`unwrap` would have been a better name for it before)
- It's name suggests `List[List[reflection_table]] -> List[reflection_table]` - In no universe do I expect a function called `flatten_reflections` to internally manipulate my data. For a `List[reflection table]` this is a little unclear - maybe, merge them? (but that'd be `merge_reflection_tables`...)
- At the very least this should be a separate, explicit function call in existing scripts. The fact that this is already called isn't good enough a reason to put it in here

If you want to merge tables at some point then this would need to be handled at the point of merge. Otherwise, isn't it only helping code that doesn't know it's using multiple tables, but is using multiple tables? Any code that wasn't broken already would have had to assume that the same ID on different tables could be a different experiment... I'm not sure who/what code this helps other than transforming for the sake of it.

Also, I'm not convinced that you can universally guarantee that number of identifiers in a table is exactly equal to number of experiments, in which case this transformation completely breaks code in a probably silent way. There's literally no way to validate this because the function doesn't have access to the experiment lists or experiments.

If this is was just an intermediate before deprecating the use of id to help stop things that used the id as a lookup index, then it'd be better to just transform all the ID's to > Nexperiments e.g. 100+. 

Basically, not sure what this actually helps by applying to *literally every single reflection table every imported*. It's a leaky abstraction that everywhere that might benefit from needs to be aware of what's going on anyway?

**Comment by jbeilstenedmands** [comment]
I’ve been thinking about this further and am inclined to agree with you Nick. In the expt_identifiers branch I’ve made some changes which aim to simplify and clarify the behaviour (relevant changes in ```dials.util.options``` and ```dials.util.multi_dataset_handling```).

I agree now that we shouldn’t be renumbering the ids in the tables on loading by default. I think the only thing that should be done is checking that the reflections and experiments are in the correct order. So I have made a new function, reflections_and_experiments_from_files, which does the job of flatten_reflections and flatten_experiments, and also rearranges the list of reflection tables to match the experiments order.
```
def reflections_and_experiments_from_files(
    reflection_file_object_list, experiment_file_object_list
):
    """Extract reflection tables and an experiment list from the files.

    If experiment identifiers are set, the order of the output reflection
    tables will match the order of the experiments, and not necessarily
    the input order of the reflection tables.
    """
    tables = [o.data for o in reflection_file_object_list]

    experiments = ExperimentList()
    for o in experiment_file_object_list:
        experiments.extend(o.data)

    if tables and experiments:
        tables = sort_tables_to_experiments_order(tables, experiments)

    return tables, experiments
```

Then, if you want to renumber id columns, there is a separate function you can call; ```renumber_table_id_columns``` from ```dials.util.multi_dataset_handling```.
In ```dials.util.multi_dataset_handling```, to split multi-dataset tables there also exists 
```split_reflection_tables_on_ids```, which splits a list of tables into a list of single-dataset tables (these two are wrapped together in ```parse_multiple_datasets``` which is currently used by scale, cosym etc.). I think these functions should be sufficient to perform the id wrangling in a simpler, clearer manner.

I welcome feedback on whether people think this is a sensible approach.
